Handle complexities of language naming (BL-5223 / BL-5244)

StephenMcConnel opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/bloombooks/bloomdesktop/1952)
<!-- Reviewable:end -->

JohnThomson commented:
Maybe I'm complicating things...but I wonder whether there is or should be any connection between this and the language-name-switching we do for Chinese in (Palaso) LanguageLookupControl.UseSimplifiedChinese()? Is switching Indonesia to Bahasa Indonesia something that should be done for all our programs, and if so, could something be in libpalaso to achieve it? Of course what you're changing here is probably nothing to do with LanguageLookup...but does that need changing too? Should we tweak Chinese here as we do in the language lookup control?
Even if this is bloom-only, I wonder whether you could find a place to put a common method for doing this? If someone else wants a substitute name, it would be very easy to only add it in one of the two necessary places.

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-:-KyWEKR0noUhkjYxXeNw:bfygk1g)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
I decided it would be a good idea to consolidate the code in one place, but at the same time there's another problem that urgently needs fixing that crept into this fix.  Languages that aren't recognized by Mono were essentially invisible to localization because it refused to create CultureInfo objects for them.  (like what .Net used to do).  So this fix is waiting on https://github.com/sillsdev/l10nsharp/pull/25 before it can be considered.

---

Review status: 0 of 5 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-:-KyXiaZPBdonAXjwBKUL:banw2sm)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
With all the code reorganization (some of which I know, I asked for) and now addressing two distinct problems it's very hard to review. Possibly what you have is the best that can be done without ICU. But the main new method doesn't seem to do quite what the comments say. Possibly the best that can be done is to adjust the comments to reflect what we'd like the method to do, with a note that currently it really does something less desirable?
I wonder if we actually use the ability to request a name for a language in some language other than itself or English? We don't seem to have actually implemented it, so maybe we just shouldn't have it?

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[src/BloomExe/Collection/CollectionSettings.cs, line 309 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzARRcNhksyqyrsi8hX:-KzARRcNhksyqyrsi8hY:btcbjxu) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/Collection/CollectionSettings.cs#L309)):*
> ```Smalltalk
> 		{
> 			var name = _lookupIsoCode.GetLocalizedLanguageName(code, inLanguage);
> 			// This behavior is wanted in GetLanguage1Name_NoCache(), but not here apparently.
> ```

This is probably moved from something existing, but I wish the commented out code was removed, and if helpful a comment explained WHY we don't want this behavior here that is in GetLanguage1Name_NoCache

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 87 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzANCwlaZ33mlsGIzcf:-KzANCwm1hwbxNUUopTv:bxexkfp) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L87)):*
> ```Smalltalk
> 				_mapCodeToIcuLocale.Add(icuDisplayCode, displayLocale);
> 			}
> 			name = locale.GetDisplayName(displayLocale);
> ```

I guess the code above is not actually used even on linux? Because I don't see how it can even compile...'name' is used before it is declared.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 101 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzALs-EeYbVHI2BYe5r:-KzALs-EeYbVHI2BYe5s:bebrl1s) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L101)):*
> ```Smalltalk
> 				{
> 					nativeName = ci.NativeName;
> 					var testChar = nativeName[0];
> ```

No way NativeName can be null or empty?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 119 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzAT2Fb6h_ow-9Lc-9q:-KzAT2Fb6h_ow-9Lc-9r:b-ev6p70) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L119)):*
> ```Smalltalk
> 						var idxCountry = englishName.IndexOf(", China");
> 						if (englishName.Length > 0 && idxCountry > 0)
> 							englishName = englishName.Substring(0, idxCountry) + "))";
> ```

Why two closing parens here?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 125 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzATSoX2KvY5gxAC0Jz:-KzATSoX2KvY5gxAC0K-:b-6ijkl0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L125)):*
> ```Smalltalk
> 				else
> 				{
> 					// CultureInfo.DisplayName is known to be broken in Mono.  It always returns EnglishName.
> ```

This else is, if I've matched the braces right, the case where we are not adding an english subtitle. It surprises me that the branches appear to differ by more than just adding it or not. The other branch never in any circumstances uses DisplayName, for example. Why doesn't this branch just use NativeName unconditionally, like the other branch does (I think) for the bit before the optional English name? It's not obvious that either branch is doing what the comment on the first caller above says, looking first for a name in the specified language...in fact, the uiCode argument only appears to be used for comparing with English and in diagnostics. Maybe this is doing the best approximation possible without ICU...but if so, it seems some comments shoudl say so.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 142 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzAVC1iJ10tLj50RnV9:-KzAVC1iJ10tLj50RnVA:b-ll2mfq) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L142)):*
> ```Smalltalk
> 				// ignore exception, but log on terminal.
> 				Console.WriteLine(@"LanguageLookup ignoring exception: {0}", e.Message);
> 			}
> ```

Given the complex logic above, I want a summary of the conditions we're handling here.
- Some exception was thrown
- we ended up with a name that stars with "Unknown Language"

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 162 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzAUfnwlrs_v9ved6P-:-KzAUfnwlrs_v9ved6P0:bg5cji9) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L162)):*
> ```Smalltalk
> 				if (nativeName != null)
> 					return nativeName + " (" + englishName + ")";
> 				return "(" + englishName + ")";
> ```

We don't want the code here? And do want the parens?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-:-KzAQBDaMpOVTDZdFhSj:b382mqw)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
I've reorganized the code in this file in an attempt to make it less confusing by splitting up the one big private method.  This results in a few lines of code and a comment or two being duplicated, but maybe it is better overall.   I hope it doesn't feel too much like a moving target.

---

Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/BloomExe/Collection/CollectionSettings.cs, line 309 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzARRcNhksyqyrsi8hX:-KzAwjrdewH1tKq7rjKb:b1bkbhh) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/Collection/CollectionSettings.cs#L309)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This is probably moved from something existing, but I wish the commented out code was removed, and if helpful a comment explained WHY we don't want this behavior here that is in GetLanguage1Name_NoCache
</blockquote></details>

I've removed the commented out code.  A unit test was enforcing the different behavior.   I guess this wasn't a L1necessarily.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 87 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzANCwlaZ33mlsGIzcf:-KzAxbHqVljgHoTYd4_U:br45gvi) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L87)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I guess the code above is not actually used even on linux? Because I don't see how it can even compile...'name' is used before it is declared.
</blockquote></details>

I've added a comment confirming the USING_ICU code is not used even on Linux, and changed the last line to be a return statement since I think that's what it would be in context if it were used.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 101 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzALs-EeYbVHI2BYe5r:-KzAzNi9nbqVdsve2l0m:bpgrsbr) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L101)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

No way NativeName can be null or empty?
</blockquote></details>

good point.  I didn't think so, but who knows?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 119 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzAT2Fb6h_ow-9Lc-9q:-KzBGkGdzB5piSYrk_Hi:b-fvo4fp) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L119)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Why two closing parens here?
</blockquote></details>

to close both Chinese and Simplified

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 142 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzAVC1iJ10tLj50RnV9:-KzBGrqMsDfcN1GKeqIy:b-yl5jip) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L142)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Given the complex logic above, I want a summary of the conditions we're handling here.
- Some exception was thrown
- we ended up with a name that stars with "Unknown Language"
</blockquote></details>

I've added a comment about this.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 162 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzAUfnwlrs_v9ved6P-:-KzB9Egco312aLdsR8NQ:b-e85138) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L162)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

We don't want the code here? And do want the parens?
</blockquote></details>

good point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-:-KzB9HcdVYDU4j-7x46a:b-gyfd7g)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
Much improved, but I caught a few new things that might deserve some attention.

---

Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 89 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOrsKMddtXOsEC_rTi:-KzOrsKMddtXOsEC_rTj:bn5d9y0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L89)):*
> ```Smalltalk
> 				// installed as a Spanish language, French language, or whatever language, system.
> 				var name = ci.DisplayName;
> 				if (name == ci.EnglishName && generalUiCode != "en")
> ```

So this amounts to detecting that DisplayName is not doing what we want? And if we can't get the language name in uiCode we think it's better to have a 'native' name than an English one, even if we're not displaying the English name in parens?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 107 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOsi_ATP5w4bSNbgFc:-KzOsi_ATP5w4bSNbgFd:bvnmfm9) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L107)):*
> ```Smalltalk
> 			// and that are being localized for Bloom.
> 			var nativeName = GetBestNativeName(code);
> 			if (generalUiCode == "en" || nativeName == null)
> ```

worth doing string.IsNullOrWhitespace here on nativeName? Or do you know for sure that null is the only no-useful-data case?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 131 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOtwC1itFwJSiE7rbx:-KzOtwC1itFwJSiE7rby:b-5845e3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L131)):*
> ```Smalltalk
> 				var testChar = nativeName[0];
> 				if ((ci.EnglishName != ci.NativeName && !IsLatinChar(testChar)) || String.IsNullOrEmpty(ci.NativeName))
> 					englishNameSuffix = " (" + ci.EnglishName + ")";
> ```

The decision tree is getting quite complex here, but in other branches we tested for not having an English name, fixed botched ones, filled in a few missing ones...should we use that logic here too?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 136 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOuZhOq64v8FizbG3O:-KzOuZhOq64v8FizbG3P:bvmjb0b) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L136)):*
> ```Smalltalk
> 				{
> 					var idxCountry = englishNameSuffix.IndexOf(" (");
> 					if (englishNameSuffix.Length > 0 && idxCountry > 0)
> ```

Is this ever going to be true? EnglishName will normally begin with " (", so idxCountry will be zero, even if there's c country code. Don't we need to start the search at index 1, or use lastIndexOf?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 152 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOvmTCLveCGTjv9okM:-KzOvmTCLveCGTjv9okN:b9u9ht7) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L152)):*
> ```Smalltalk
> 				}
> 				var name = nativeName + englishNameSuffix;
> 				name = FixBotchedName(name);
> ```

I suppose this will work because the only current case for FixBotchedName is a latin-script original that won't get an English suffix. But it would be more robust to fix it before adding the suffix, wouldn't it?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 216 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzP0iPttq2POmW0dtuj:-KzP0iPttq2POmW0dtuk:b-f2nb1u) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L216)):*
> ```Smalltalk
> 
> 		/// <summary>
> 		/// For the languages we know about, return the native name.  If we don't know anything, return null.
> ```

Seems the method name, or at least the comment, should indicate that this is only a few exceptions...it doesn't work for the many languages where CultureInfo has a NativeName.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-:-KzP1B7UHPTdCozzUGBD:b9iq2xo)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 125 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzATSoX2KvY5gxAC0Jz:-KzZvH2r64CemkmWtDVO:byfppip) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L125)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

This else is, if I've matched the braces right, the case where we are not adding an english subtitle. It surprises me that the branches appear to differ by more than just adding it or not. The other branch never in any circumstances uses DisplayName, for example. Why doesn't this branch just use NativeName unconditionally, like the other branch does (I think) for the bit before the optional English name? It's not obvious that either branch is doing what the comment on the first caller above says, looking first for a name in the specified language...in fact, the uiCode argument only appears to be used for comparing with English and in diagnostics. Maybe this is doing the best approximation possible without ICU...but if so, it seems some comments shoudl say so.
</blockquote></details>

Adding the English suffix or not is done with a single line if conditioned statement.  I've tried to simplify the logic by putting the condition test into a separate method.  That test doesn't (and didn't) need any braces or else.  The braces and else are for handling Chinese where we want to keep "Chinese (Simplified)" but not "Chinese (Simplified, China)" which is minimally what .Net produces.  (I've seen something like "Chinese (Simplified, China, International Sort)" or something like that once or twice for some reason.)  And this method never uses DisplayName.  It does use Name, which is CultureInfo's confusingly named property for the ISO code to test for Chinese.

I'm sorry you've found this code so confusing.  Dealing with an area where one system throws exceptions and another system fills in empty objects for a class that neither system will allow you to create and initialize yourself is inherently a bit complex.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 89 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOrsKMddtXOsEC_rTi:-KzZYDA7bUKU7cTwWCqy:bplwvd3) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L89)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

So this amounts to detecting that DisplayName is not doing what we want? And if we can't get the language name in uiCode we think it's better to have a 'native' name than an English one, even if we're not displaying the English name in parens?
</blockquote></details>

Added more commentary to this block of code making this deduction explicit.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 107 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOsi_ATP5w4bSNbgFc:-KzZYBJXSjjHJaMB0vKv:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L107)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

worth doing string.IsNullOrWhitespace here on nativeName? Or do you know for sure that null is the only no-useful-data case?
</blockquote></details>

Done.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 131 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOtwC1itFwJSiE7rbx:-KzZsA0nvjkwRfkQuSEz:b-rsq1wj) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L131)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

The decision tree is getting quite complex here, but in other branches we tested for not having an English name, fixed botched ones, filled in a few missing ones...should we use that logic here too?
</blockquote></details>

I've renamed the method to FixBotchedNativeName and applied to the final native name we come up with in this method.  I also changed the logic to apply this method in the initial assignment of ci.NativeName to a local variable in both functions that call it.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 136 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOuZhOq64v8FizbG3O:-KzZqDT_XMsBOTriZkVG:bw4ab1d) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L136)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Is this ever going to be true? EnglishName will normally begin with " (", so idxCountry will be zero, even if there's c country code. Don't we need to start the search at index 1, or use lastIndexOf?
</blockquote></details>

Good catch.  It should be LastIndexOf(" (") not IndexOf(" (").

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 152 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzOvmTCLveCGTjv9okM:-KzZtefIFtWMhdt65oJe:b-896fix) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L152)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I suppose this will work because the only current case for FixBotchedName is a latin-script original that won't get an English suffix. But it would be more robust to fix it before adding the suffix, wouldn't it?
</blockquote></details>

Done.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 216 at r3](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzP0iPttq2POmW0dtuj:-KzZaC_f8sX5jEXidLOf:b-r7qthn) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/a355320bf31674f06e1f8c5e115b3bfec0da6ebf/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L216)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Seems the method name, or at least the comment, should indicate that this is only a few exceptions...it doesn't work for the many languages where CultureInfo has a NativeName.
</blockquote></details>

renamed to GetNativeNameIfKnown and expanded the method comment to clarify its limited utility/knowledge.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
I've renamed a few things, added caching of language names (which was there at one point), and rearranged the code a bit.  I hope this is satisfactory now.

---

Review status: 3 of 5 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-:-Kz_9VkyQEYDHOIgymSm:b-b9jvp1)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
I'm much happier with the organization of the code. Thanks for your patience. Something prompted me to try CultureInfo.GetCultureInfo on a few languages, and the results on Windows are alarmingly poor...see my comment below. I also had one other question, which could perhaps reasonably be left for testers who hopefully have non-English OS's available to test. If neither of these seems worrying to you, or the improvement is critical enough that it makes sense to address them later, I think you could go ahead and merge.

---

Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 125 at r2](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-KzATSoX2KvY5gxAC0Jz:-Kz_GKnNEBZZsOOLK78J:br9npcl) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/302f000ede3f7579b392bd7d98b913ccda6d2bb3/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L125)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

Adding the English suffix or not is done with a single line if conditioned statement.  I've tried to simplify the logic by putting the condition test into a separate method.  That test doesn't (and didn't) need any braces or else.  The braces and else are for handling Chinese where we want to keep "Chinese (Simplified)" but not "Chinese (Simplified, China)" which is minimally what .Net produces.  (I've seen something like "Chinese (Simplified, China, International Sort)" or something like that once or twice for some reason.)  And this method never uses DisplayName.  It does use Name, which is CultureInfo's confusingly named property for the ISO code to test for Chinese.

I'm sorry you've found this code so confusing.  Dealing with an area where one system throws exceptions and another system fills in empty objects for a class that neither system will allow you to create and initialize yourself is inherently a bit complex.
</blockquote></details>

I'm sorry, too. At some points I may have been being too picky, but I think the code has really improved.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 173 at r4](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-Kz_G52J0pJkyBfFQ-p5:-Kz_G52J0pJkyBfFQ-p6:b-bca9zw) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/96e032e084e578382c9b0df775ef276bbe14c638/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L173)):*
> ```Smalltalk
> 				}
> 				langName = nativeName + englishNameSuffix;
> 				if (!langName.StartsWith("Unknown Language"))
> ```

Sorry...just occurred to me. Will "unknown language" be localized to something else on one or other platform if English is not the main OS language?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 277 at r4](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-Kz_Gs7hf9qLTdON62HZ:-Kz_Gs7hf9qLTdON62H_:bie3bnx) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/96e032e084e578382c9b0df775ef276bbe14c638/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L277)):*
> ```Smalltalk
> 			{
> 			case "pbu":  return "پښتو";
> 			case "prs":  return "دری";
> ```

I'm surprised there are so few. Of the languages Bloom is already localized into, CultureInfo.GetCultureInfo() on Windows doesn't know about Tok Pisin, Hausa, Khmer, or Tamil...I just tried four of the more major languages that aren't obviously big enough to have 2-letter codes; probably several others don't work. I'm not sure where these names are used, but it would be unfortunate if our UI language menu deteriorated from showing "Tamil" to "tam"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-:-Kz_KUfvbeHUNhwXDK0y:b-meazrz)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 173 at r4](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-Kz_G52J0pJkyBfFQ-p5:-Kz_Oy521-AL_7zqYxmh:b-86hi4p) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/96e032e084e578382c9b0df775ef276bbe14c638/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L173)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

Sorry...just occurred to me. Will "unknown language" be localized to something else on one or other platform if English is not the main OS language?
</blockquote></details>

hmm, I suppose that depends on whether we're looking at EnglishName or DisplayName or NativeName.  This is Windows-only behavior since Mono throws an exception for languages it doesn't know about.  (I think Windows .Net used to do that.)  I'll check ci.EnglishName explicitly for this test.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 277 at r4](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-Kz_Gs7hf9qLTdON62HZ:-Kz_Qf6p4Z4iURfUQlpa:b-njw1x0) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/96e032e084e578382c9b0df775ef276bbe14c638/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L277)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary><blockquote>

I'm surprised there are so few. Of the languages Bloom is already localized into, CultureInfo.GetCultureInfo() on Windows doesn't know about Tok Pisin, Hausa, Khmer, or Tamil...I just tried four of the more major languages that aren't obviously big enough to have 2-letter codes; probably several others don't work. I'm not sure where these names are used, but it would be unfortunate if our UI language menu deteriorated from showing "Tamil" to "tam"
</blockquote></details>

I see Hausa, Khmer, and Tamil listed in the menu properly already on Linux and Windows in the language selection menu.  (Apparently they do have 2-letter codes.)  I don't see Tok Pisin at all.  When I add a dummy tpi/Bloom.xlf file, it comes up as "tpi (Tok Pisin)" on Linux and "Unknown Language" on Windows.  Adding this language to the unknown language handling causes "Tok Pisin" alone to show up in the menu on Linux (and presumably on Windows).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952)*
<!-- Sent from Reviewable.io -->

JohnThomson commented:
I don't really know where this is used. In general, I think it is ready to be merged if it's only used for names of wider-communication languages, especially if we control which ones. If it's used for vernacular ones, I wonder whether just handling those that have 2-letter codes is good enough. But maybe it's the best we can reasonably do without the overhead of ICU? Maybe discuss after standup? Sorry to keep holding things up.

---

Reviewed 1 of 2 files at r3, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 173 at r4](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-Kz_G52J0pJkyBfFQ-p5:-KzxsyR443pXtkL895Z5:b7dvlpx) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/96e032e084e578382c9b0df775ef276bbe14c638/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L173)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

hmm, I suppose that depends on whether we're looking at EnglishName or DisplayName or NativeName.  This is Windows-only behavior since Mono throws an exception for languages it doesn't know about.  (I think Windows .Net used to do that.)  I'll check ci.EnglishName explicitly for this test.
</blockquote></details>

Hmm, if I was writing lookup code that is supposed to produce an English language name for a French-speaking user and was thinking about how to handle an unknown code, I would be tossing up between returning "Unknown Language" or the French equivalent. Maybe this doesn't matter much, if it is only used for lists of languages we control like the list of UI languages, and in practice if this result appears we will just extend our list of special cases? But if it's used anywhere to find names of vernacular languages, I think we should actually find out whether it works in a French-UI Windows OS. Maybe the easiest way to do that is to merge it and let Len try it out?

---

*[src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs, line 277 at r4](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-Kz_Gs7hf9qLTdON62HZ:-Kzxu9iSAK0KFRiFzeX9:b9rw4wb) ([raw file](https://github.com/bloombooks/bloomdesktop/blob/96e032e084e578382c9b0df775ef276bbe14c638/src/BloomExe/ToPalaso/LanguageLookupModelExtensions.cs#L277)):*
<details><summary><i>Previously, StephenMcConnel (Steve McConnel) wrote…</i></summary><blockquote>

I see Hausa, Khmer, and Tamil listed in the menu properly already on Linux and Windows in the language selection menu.  (Apparently they do have 2-letter codes.)  I don't see Tok Pisin at all.  When I add a dummy tpi/Bloom.xlf file, it comes up as "tpi (Tok Pisin)" on Linux and "Unknown Language" on Windows.  Adding this language to the unknown language handling causes "Tok Pisin" alone to show up in the menu on Linux (and presumably on Windows).
</blockquote></details>

If that's all it's used for, I guess we're fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomdesktop/1952#-:-KzxuKd1R6QLxAGM6pGU:b-8kqi9x)*
<!-- Sent from Reviewable.io -->
